Specify the symbol file format (#37)
gbdev/rgbds-www: src/pages/sym.mdx

---
title: Game Boy symbol format specification
---

import Link from '@docusaurus/Link';
import { latestStable } from '@site/src/components/versions.js';

export const LatestStableLink = ({children, to}) => <Link to={`/docs/${latestStable}/${to}`}>{children}</Link>;

# Game Boy symbol (`.sym`) file format specification

## Purpose & Context

> Game Boy ROM files contain pure binary data, but no metadata.

Author: I tried to say that Game Boy ROMs are raw binary dumps with no extra info, including a symbol table.

This document aims to standardize the format and usage of sym files in order to avoid this problem.
Modifications, etc. should be suggested by [opening an issue](https://github.com/gbdev/rgbds-www/issues) or [starting a discussion](https://github.com/gbdev/rgbds-www/discussions) on this website's source repository, whichever is more appropriate.

Extensions to this format are known to exist ([example](https://github.com/mattcurrie/mgbdis/#symbol-files)), but they tend to be specific to a certain use case, and are generally a backwards-compatible superset of what are described below.
Keep in mind that some sym files are written manually, such as when disassembling ROMs, so it is recommended for consumers to be lenient in what they accept.

Note also that [wla-dx's sym files](https://wla-dx.readthedocs.io/en/latest/symbols.html) are much more complex (and complete) than this format.
RGBDS aims to keep sym files simple, and instead chooses to output other data in different files, such as map files[^1].

> [^1]: Map files remain thus far unspecified. Very few tools rely on information contained in them, so no specification has been written in order to keep that format extensible.

Author: This footnote is there to give the rationale as to why they are unspecified, and to point out that relying on them *is* fragile until then.

I know you have such tooling, but as the footnote points out, this is a tiny minority. If that number were to grow, a spec would be written.

## File format specification

:::info

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

Reviewer: I think this would be too distracting to read. The formalism isn't worth it in a doc for such a simple format, for a hobbyist tool, where implementers could probably guess most of the spec details anyway without a formal doc.

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

Reviewer 2: Maybe I'm used to it? Is it really that distracting to have a capitalized word?

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

Author: I do find it so.

> The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

Author: Note that RFC 8174 also updates RFC 2119 to restrict its scope to IETF documents, so using the original version makes more sense, if only pedantically.

Numbers are written in decimal by default, or hexadecimal if prefixed with a `$`, unless otherwise specified.

:::

### Tokenization

> Line endings can be either LF-only or CRLF, though the former is preferred; they need not be consistent throughout the document.

Reviewer: It would be more confusing to have to view two independent sections just to know what to write on some `if (c == '\r' || c == '\n')` line.

> Line endings can be either LF-only or CRLF, though the former is preferred; they need not be consistent throughout the document.

Reviewer 2: That's exactly my point. All it should say here is that line endings can be either, so you know you have to check for both.

> Line endings can be either LF-only or CRLF, though the former is preferred; they need not be consistent throughout the document.

Reviewer: How about: `` Line endings may be either LF-only (`\n` U+000A) or CR+LF (`\r\n` U+000D U+000A).``

> Line endings can be either LF-only or CRLF, though the former is preferred; they need not be consistent throughout the document.

Reviewer 2: Perfect!

> The character encoding must be well-formed UTF-8 (without BOM), however it is strongly encouraged to stick to ASCII characters for portability.

Author: Having to bounce between two sections more than a screenful apart that both touch the same topic is more confusing, so, no.

> The character encoding must be well-formed UTF-8 (without BOM), however it is strongly encouraged to stick to ASCII characters for portability.

Reviewer 2: This one in particular I don't feel too strong about. It's clearly a recommendation remark, and for good reason, so it's not interfering with anything.
The requirements for encoders and decoders are clear, anyway: encoders should prefer ASCII, but may use UTF-8; decoders must support UTF-8, but can favour ASCII.

Especially since they are intended to be somewhat human-readable, sym files can contain comments.
Comments begin with a semicolon (`;`, U+003B) anywhere in a line, all the way to the end of that line.
Semicolons cannot be escaped, as they are not valid in any part of entries.
Comments should be ignored by the implementation.

After removing the comment (if any), the line shall be split into non-empty **tokens**, separated by one or more consecutive **whitespace** characters.

> ASCII spaces (` `, U+0020) must be treated as whitespace; implementations may also treat tabs (`\t`, U+0009) as whitespace, but this is not portable to some older implementations.

Author: Again, traditional implementations don't support spaces, and that's what you should use for portability; however, many humans would end up using tabs because that's practical to them or w/e, and they should be able to if their emulator supports them. This is also why tabs are *always* excluded from symbol names below.

> ASCII spaces (` `, U+0020) must be treated as whitespace; implementations may also treat tabs (`\t`, U+0009) as whitespace, but this is not portable to some older implementations.

Reviewer 2: That's why I say this is the worst of both words. Just mandate spaces, then! If an implementation wants to extend the format by supporting something invalid, that's fine; that's what we call a format extension. But the specification doesn't need that.

Since a file with tabs in it is invalid by the spec, a strict implementation would error out on it. Let implementations silently handle those files however they wish; don't muddy up the spec. Instead, teach those humans to use spaces instead of tabs, the same way we teach them to not do pointer arithmetic on `void *` in C even though their compiler supports it.

> ASCII spaces (` `, U+0020) must be treated as whitespace; implementations may also treat tabs (`\t`, U+0009) as whitespace, but this is not portable to some older implementations.

Author: The rest of the spec is written to allow tabs as whitespace (e.g. making them part of the character set excluded from tokens), so it wouldn't make sense to allow spaces only. But requiring tabs makes older implementations non-conformant, and I want to avoid friction. Hence this.

> ASCII spaces (` `, U+0020) must be treated as whitespace; implementations may also treat tabs (`\t`, U+0009) as whitespace, but this is not portable to some older implementations.

Reviewer 2: Making older implementations non-conformant is fine. They can update. The ones that can't will remain non-conformant, and that's fine, because they didn't have time machines and we don't expect a tool from 2004 to conform to a specification written in 2022.
Old, unupdatable implementations will always fall behind the state of the art. That's normal, and it's the users' choice to use such an implementation, so long as they are aware of this difference. Considering that this is going to be used by the same community that offers SDCC/GBDK as a toolchain, a C compiler that is only half conformant to C89 and doesn't even attempt to conform to later standards, I'd say this isn't inherently a problem.

> ASCII spaces (` `, U+0020) must be treated as whitespace; implementations may also treat tabs (`\t`, U+0009) as whitespace, but this is not portable to some older implementations.

Author: beware's gonna hate me, but you have a point. I'll make tabs a requirement.

Other Unicode "White Space" characters must *not* be treated as whitespace.
Leading or trailing whitespace must be ignored.

### Line processing

> If any part of a line cannot be processed by the implementation, it (and possibly the whole line) shall be ignored, and a warning should be produced.

Author: The goal was to specify that a line that cannot be fully processed shouldn't be rejected outright, and that an invalid line does not invalidate the entire file.

The rationale for the latter should be obvious, whereas the rationale for the former is about some amount of backwards compatibility, e.g. any tokens after a symbol definition (which may be later defined to carry extra info, such as file provenance, memory type, etc.) should not invalidate the symbol definition if the implementation does not recognize them.

A line without any tokens shall be silently ignored, to permit empty lines, even if they contain whitespace and/or a comment.

> A line with only one token shall be ignored, being reserved for future extensions.

Author: Such extra tokens would be used to provide additional information about the symbol, but the base info would still be available. Lines with entirely different meaning would use syntax outright different from the symbol definitions.

> A line with only one token shall be ignored, being reserved for future extensions.

Reviewer 2: You're giving yourself no room for extensions. You already define lines with zero tokens (blanks) and with two or more (symbols). If you also define lines with one token, that defines _every possible line you could write_. Literally every line has zero tokens, one token, or two or more tokens.

> A line with only one token shall be ignored, being reserved for future extensions.

Author: The room for extensions is in the first token: currently, it has to be a valid address specification, but new meaning can be assigned later to when it is not.

> A line with only one token shall be ignored, being reserved for future extensions.

Reviewer 2: I'd add a note where address formats are listed, then, stating that other forms are reserved for future extension if necessary.

Encountering one may produce a warning.

Otherwise, the line defines a single **symbol**: the first token provides its location, the second token provides its name, and further tokens (if any) may provide extra metadata.
Extra tokens are currently not specified, but may be so in the future; so, implementations should ignore them, and may produce a warning in doing so.

#### Symbols

A symbol is uniquely defined by the combination of its location and its name; symbol names are case-sensitive.

> If the same symbol (identical name *and* identical location) is defined twice, one of the two definitions must be silently ignored.

Author: Given that no metadata is defined, that's up to the implementation.

> If the same symbol (identical name *and* identical location) is defined twice, one of the two definitions must be silently ignored.

Reviewer 2: That leaves a spec gap that shouldn't be there. I'd explicitly write that metadata must be identical or behavior is undefined.

> If the same symbol (identical name *and* identical location) is defined twice, one of the two definitions must be silently ignored.

Reviewer: There *is* no metadata. If a future spec version adds any, the spec could be changed.

A warning may be produced, though not necessarily when loading the sym file: for example, "displaying names instead of memory addresses" accomodates this case just fine, however a "go to symbol" feature wouldn't, and a warning could be produced when the user tries to go to an ambiguous target.

##### Location

The location shall be of the form `<bank>:<address>`, both `<bank>` and `<address>` being case-insensitive hexadecimal numbers, possibly zero-padded, separated by a colon (`:`, U+003A).
Typical hexadecimal prefixes (`0x`, `$`, etc.) are not allowed.
An implementation may also allow the bank to be the case-insensitive keyword `BOOT` to specify symbols in the loaded boot ROM.
If the bank is omitted, then the colon must be as well; the results of such lines are implementation-defined.

The address must be in 16-bit range (0 to $FFFF inclusive), the bank must be in 32-bit range (0 to $FFFFFFFF inclusive); implementations may abandon processing the line if either number is larger than these limits.
Implementations may accept bank numbers higher than $FFFFFFFF.

> Nonsensical locations (e.g. `2:9800`) may produce a warning.

Reviewer 2: Don't hedge on the boot bank (just allow it). Also, don't mistake supporting an address format for accepting the symbol definition.
I'd honestly rewrite the whole thing. There's so many fuzzy requirements here that nobody could possibly hope to know what's valid and what's not. Let me be clear about something: the whole point of a specification is to define behaviors. If existing behaviors are inconsistent, they'll have to adapt. There's _no way_ to write a specification that is useful and accomodates all existing implementations from the past 20+ years; any attempt at the former will have to forgo the latter to some extent. Existing implementations can adapt if the changes are minimal.

"Locations can have one of the following three forms:

- `<number>:<number>`, indicating a location by its bank and address;
- `BOOT:<number>`, indicating a location in the boot ROM (the GB's internal monitor ROM or any emulator stand-in); or
- `<number>`, indicating a location by its address irrespective of bank (i.e., a location in every possible bank).

All forms are case-insensitive; in all cases, `<number>` is a sequence of one to four hexadecimal digits, without any prefixes indicating the base (like `0x` or `$`). Implementations MAY ignore symbols defined with locations in the last two forms if they don't support those types of addresses. Implementations MAY issue a warning if a location is nonsensical for the chosen hardware configuration, such as  `2:9800` (or `1:9800` in DMG mode)."

Note that I removed banks greater than 16 bits. There's exactly zero use cases for them. If some implementation supports garbage there, great, but there's no point putting it on a spec.

Two locations are considered identical if they have the same bank and the same address.
Two distinct symbols can have identical locations; in particular, <LatestStableLink to="rgbasm.5#Unions">unions</LatestStableLink> facilitate this.

> Locations are strictly ordered by memory region (ROM0, ROMX, VRAM, SRAM, WRAM0, WRAMX, Echo RAM, OAM, I/O, HRAM, IE), then by bank, then by address.

Reviewer 2: What does this mean in terms of requirements? What part of the specification cares about this?

Entries may or may not be sorted in the sym file; for example, RGBLINK before v0.4.0 output symbols in the same order that it read them from its input files.

> Some sym files are piped through [`sort(1)`](https://man.openbsd.org/sort.1), which often produces files sorted by bank then address, but not by region.

Reviewer 2: And sometimes the sun is up, and sometimes it's covered by clouds, and sometimes it's night time.
This specifies nothing. Leave it out. If you really must, a simple "Entries may appear in any order" is enough; nothing else is being specified here.

Sym files do not carry information about the memory region a symbol belongs to, but that can be inferred from the address most of the time.

> This detection is sometimes ambiguous: for example, `0:A000` could be one past the end of VRAM, or at the beginning of SRAM.

Reviewer 2: Everything from here until the end of the paragraph should migrate to a non-normative section.

Reviewer: The should/may/must/etc keywords are already sufficient to tell what's required versus normative. No need to spread related information across distant sections for that.

Reviewer 2: The rest of the information isn't even related; it's just a recommendation about how to assign symbols to memory regions, which isn't part of the specification at all! Unless this has direct relevance to the specification (in which case, it should be a must — and it's a _terrible_ idea), I'd be happy to see it deleted entirely. But since someone might find the recommendations useful, moving them to the bottom seems like a reasonable compromise.

Author: Recommendations far away from the thing they refer are worse, since having to jump around in the document is terrible for comprehension. This is why it's here.

These paragraphs can be made into "boxes" to make their non-normative status clearer.

Reviewer 2: This one case in particular is significant because (1) it's talking about heuristics, (2) it's introducing a lot of novel behavior that isn't related to sym file parsing, and (3) it brings in a whole lot of problems that are neither useful nor necessary arising from depending on the textual meaning of labels (e.g., what if the symbol names aren't in English?).
It's a huge tangent that seems completely out of place and adds nothing to the spec at all.

Other smaller notes are fine like that, and sometimes they don't even need a box (a parenthetical is fine), but this one case is particularly egregious. I'd really like to see it simply deleted. But if you think it's important, it really doesn't belong in the main body of the specification.

The following heuristics are suggested:
- If the symbol's bank is in range of exactly one of the candidate regions, assign it to that one. (*There should be no ambiguity, e.g. `2:A000`.*)
- If the symbol is local and is "attached" to a global symbol (see the "presentation shortening" below), assign it to the same region as the global symbol. (*Local labels should "follow" the global symbol they are attached to, if any.*)
- If the symbol's name ends with (case-insensitive) "end", assign it to the earlier region. (*This is a common naming convention, and "one-past-end" labels are typically "end markers".*)
- If all else fails, assign it to the later region (e.g. `0:A000` would be assigned to RAM). (*This is correct more often than not in practice.*)

##### Name

> The symbol name can contain any characters except ASCII spaces, tabs, non-printable ASCII characters, or characters excluded by earlier processing (e.g. semicolons).

Reviewer 2: What about Unicode garbage? I'd be explicit in the ranges supported.
(And don't even attempt to find a reasonable subset of Unicode. The Unicode Consortium will ruin your work almost instantly as soon as they add a random character you don't expect.)

Author: The file format is specified to be "well-formed UTF-8", so that's already excluded.

Reviewer 2: I didn't mean invalid sequences (like a 0xc0), but rather lots of Unicode things that aren't really printable.

Author: I'm not sure what you want, given that you both say "I'd be explicit in what's supported" and "don't try to be explicit because updates will bite you".

Reviewer 2: My point is that you basically have three ways to handle Unicode characters that will survive an update cycle: accept them all as identifiers (essentially treating them as letters), reject them all, or have an explicit, non-changing whitelist/blacklist (for instance, JSON only accepts a few explicitly designated codepoints as whitespace).

While your current wording would imply the former (Unicode whitespace is a valid symbol character), I'm not too sure if this is what was intended.

Author: I'm not sure what the "update cycle" is, but my intention is that such characters should be treated as part of identifiers; whether they can be processed and how is left up to the implementation. ASCII is already encouraged as per the above, so I think that's reasonable.

Reviewer 2: In that case I'd make it explicit that excluded characters are codepoints U+0000 through U+0020. (You can add a note indicating what those codepoints are, but it's just C0 controls (including tabs) and your good old common space.)

Author: Well, that and U+007F.

Reviewer 2: I think you just proved why an explicit list is important :P

A symbol whose name does not contain any periods (`.`, U+002E) is a **global symbol**.
If a symbol's name contains exactly one period, it is a **local symbol**; the portion up to but not including the period is the **global portion**, the rest is the **local portion**; neither shall be empty.

> A symbol with two or more periods in its name shall be processed in an implementation-defined manner, which may include rejecting or ignoring it.

Reviewer 2: Define this. No reason not to.

Author: Sure, what should it be defined to, then?

Reviewer 2: Whatever you want RGBASM to do with such a symbol in the future, even if it's not implemented today.

Author: ...which involves a lot of moving parts that I don't want to design ahead of time.

A local symbol may be **attached** to a global symbol.

> If the previous (by the ordering defined above for locations) global symbol exists, and its name matches the local symbol's global portion, then it is considered attached to it.

Reviewer 2: This definition of "previous" is unclear, because it depends on memory region placement, which is not only unspecified, but _unspecifiable_ (there's no way for the sym file to contain that information in the way it is defined). Depending on random heuristics like the name of the label to define this in a normative way is the single worst idea presented in this document by a wide margin.

I'd suggest a simpler approach: a local symbol is attached to a global symbol if there exists a global symbol with the same location type (see the three location types I defined above), lower or equal address, and, if the symbol is a banked symbol, same bank. Yes, this may cause false positives (such as `2:C456 Foo.bar` being attached to `2:4567 Foo`), but honestly, if someone defines symbols like this, they deserve what they get.

> Attached local symbols may have their names shortened in presentation to just the local portion (i.e., the period and everything that follows).

Reviewer 2: Whether an interactive debugger writes out local symbols in full, shortened, with an arrow pointing to a reference, chocolate-covered or translated to ancient Phoenician doesn't concern this specification: it's not part of the file format. Just remove this.

Reviewer: In that case, the very notion of "attachment" is not part of the file format. It has no effect on parsing, only on presentation. And implementations may choose to specially display `Foo` and `EndFoo`, or `Foos` and `Foo1`, `Foo2`, `Foo3`, etc, or `Foo` and `_Foo`, or `Foo` and `FooHelper`, regardless of which one comes first.

People would read this doc to know how to implement symbol files, and the `Global.local` convention is information they'd want to know.

Reviewer 2: Sure, but that's already explained. Note that my comment was only about line 111, not the previous lines.

RGBDS aims to keep sym files simple, and instead chooses to output other data in different files, such as map files[^1].

[^1]: Map files remain thus far unspecified. Very few tools rely on information contained in them, so no specification has been written in order to keep that format extensible.

## File format specification

:::info

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119).

Numbers are written in decimal by default, or hexadecimal if prefixed with a `$`, unless otherwise specified.

:::

### Tokenization

Line endings can be either LF-only or CRLF, though the former is preferred; they need not be consistent throughout the document.
The character encoding must be well-formed UTF-8 (without BOM), however it is strongly encouraged to stick to ASCII characters for portability.
The NUL character (`\0`, U+0000) shall not appear anywhere in the file.

> An implementation encountering one must output a diagnostic if it can, and may then ignore it, the line, the whole file, or reject the file outright.

Reviewer: I don't think this needs to be specifically addressed. It's covered by these two lines:

- "If any part of a line cannot be processed by the implementation, it (and possibly the whole line) shall be ignored, and a warning should be produced."
- "The symbol name can contain any characters except ASCII spaces, tabs, non-printable ASCII characters, or characters excluded by earlier processing (~~semicolons, NULs, etc.~~ *e.g. semicolons*)."

The details of line processing, such as the definition of how the "location" is formatted, implicitly disallow NUL bytes; and  they're disallowed for symbol names because NUL is one of the non-printable ASCII characters.

Seeing the spec go into such low-level detail makes it sound like a more complicated format than it is. You could parse a standard-compliant file with a regex or two, or handle noncompliant ones with a `try { regex parse line } catch { ignore and warn }`.

Especially since they are intended to be somewhat human-readable, sym files can contain comments.

> Comments begin with a semicolon anywhere in a line, all the way to the end of that line.

Reviewer: ```suggestion
Comments begin with a semicolon (`;`, U+003B) anywhere in a line, all the way to the end of that line.
```

A line with only one token shall be ignored, being reserved for future extensions.
Encountering one may produce a warning.

Otherwise, the line defines a single **symbol**: the first token provides its location, the second token provides its name, and further tokens (if any) may provide extra metadata.
Extra tokens are currently not specified, but may be so in the future; so, implementations should ignore them, and may produce a warning in doing so.

#### Symbols

A symbol is uniquely defined by the combination of its location and its name; symbol names are case-sensitive.
If the same symbol (identical name *and* identical location) is defined twice, one of the two definitions must be silently ignored.

If two distinct symbols are defined with the same name (but different locations), neither definition shall be ignored, and this should not be an error.
A warning may be produced, though not necessarily when loading the sym file: for example, "displaying names instead of memory addresses" accomodates this case just fine, however a "go to symbol" feature wouldn't, and a warning could be produced when the user tries to go to an ambiguous target.

##### Location

> The location shall be of the form `<bank>:<address>`, both `<bank>` and `<address>` being case-insensitive hexadecimal numbers, possibly zero-padded, but without any prefix (so `$02:4000` is not a valid location, for example).

Reviewer: ```suggestion
The location shall be of the form `<bank>:<address>`, both `<bank>` and `<address>` being case-insensitive hexadecimal digits, possibly zero-padded, separated by a colon (`;` U+0058). Typical hexadecimal prefixes (`0x`, `$`, etc) are not allowed.
```

Author: It'd have to be "composed of" to use "digits".

Reviewer: Edited.

Implementations may accept bank numbers higher than $FFFFFFFF.
Nonsensical locations (e.g. `2:9800`) may produce a warning.

Two locations are considered identical if they have the same bank and the same address.
Two distinct symbols can have identical locations; in particular, <LatestStableLink to="rgbasm.5#Unions">unions</LatestStableLink> facilitate this.

Locations are strictly ordered by memory region (ROM0, ROMX, VRAM, SRAM, WRAM0, WRAMX, Echo RAM, OAM, I/O, HRAM, IE), then by bank, then by address.
Entries may or may not be sorted in the sym file; for example, RGBLINK before v0.4.0 output symbols in the same order that it read them from its input files.
Some sym files are piped through [`sort(1)`](https://man.openbsd.org/sort.1), which often produces files sorted by bank then address, but not by region.

Sym files do not carry information about the memory region a symbol belongs to, but that can be inferred from the address most of the time.
This detection is sometimes ambiguous: for example, `0:A000` could be one past the end of VRAM, or at the beginning of SRAM.
The following heuristics are suggested:

> - If the symbol's bank is in range of exactly one of the candidate regions, assign it to that one. (*There should be no ambiguity, e.g. `2:A000`.*)

Reviewer: How is `2:A000` ambiguous? There's no VRAM2, so it must be the start of SRAM bank 2.

Author: It is ambiguous if you only consider the address, which a naïve implementation might. As the line says, those cases actually have no ambiguity.

Reviewer: It says "There should be no ambiguity, e.g. `2:A000`." I can only read this as "There should be no ambiguity, such as this ambiguous example, `2:A000`." How about (removing the italics too): `(There should be no ambiguity, `1:8000`, which can be in ROMX or VRAM.)`

- If the symbol is local and is "attached" to a global symbol (see the "presentation shortening" below), assign it to the same region as the global symbol. (*Local labels should "follow" the global symbol they are attached to, if any.*)
- If the symbol's name ends with (case-insensitive) "end", assign it to the earlier region. (*This is a common naming convention, and "one-past-end" labels are typically "end markers".*)

> - If all else fails, assign it to the later region. (*This is correct more often than not.*)

Reviewer: ```suggestion
- If all else fails, assign it to the later region (e.g. `0:A000` would be assigned to the beginning of SRAM). (*This is correct more often than not.*)
```

Symbols defined with a location in one of the latter two forms may be ignored if such address spaces are not supported by the implementation; a warning may be emitted then.
For the first form only, a warning may be issued if the location is nonsensical for the current hardware configuration (e.g. `2:9800`).

The address must be in 16-bit range (0 to $FFFF inclusive), the bank must be in 32-bit range (0 to $FFFFFFFF inclusive); implementations may abandon processing the line if either number is larger than these limits.
Implementations may accept bank numbers higher than $FFFFFFFF.

Two locations are considered identical if they have the same bank (including none) and the same address.
Two distinct symbols can have identical locations; in particular, <LatestStableLink to="rgbasm.5#Unions">unions</LatestStableLink> facilitate this.

Sym files may or may not be sorted by location; implementations must be able to handle both.
(For example, RGBLINK before v0.4.0 output symbols in the same order that it read them from its input files.)

##### Name

> Symbol names can contain any characters except U+0000 through U+0020 (inclusive) or U+007F, or characters excluded by earlier processing (e.g. semicolons).

Reviewer: `[A-Za-z_][A-Za-z_0-9]*` is a really common cross-language symbol syntax. It's interesting that rgbasm also allows `#` and `@`, but yeah, I could imagine it or some for allowing them as leading too, and `$` is not uncommon in low-level symbol names either.

I guess applications could require `SomeLabel + 10` to be unambiguous.

> Symbol names can contain any characters except U+0000 through U+0020 (inclusive) or U+007F, or characters excluded by earlier processing (e.g. semicolons).

Author: Have an attempted compromise?

> Symbol names can contain any characters except U+0000 through U+0020 (inclusive) or U+007F, or characters excluded by earlier processing (e.g. semicolons).

Reviewer: What, like, allow any characters in symbol names, and trust users not to define `Foo+7` if they also defined `Foo` somewhere else? Sure.

> Symbol names can contain any characters except U+0000 through U+0020 (inclusive) or U+007F, or characters excluded by earlier processing (e.g. semicolons).

Author: Oh, no, I mean, what I just changed this paragraph to:

https://github.com/gbdev/rgbds-www/blob/31ee5b697ab675b080caa7d722393bbde27c29a9/src/pages/sym.mdx#L109

## Purpose & Context

Game Boy ROM files contain pure binary data, but no metadata.
In particular, symbol information is lacking when debugging or reverse-engineering such ROMs.
Symbol files, or "sym files" for short, aim to be companions to Game Boy ROM files that provide exactly that: symbol information.

Sym files have been around for a long time (the [no$gmb](http://problemkaputt.de/gmb.htm) emulator supported them, so it is safe to assume since at least the early 2000's), but the format was always loosely defined, leading to subtle incompatibilities.
This document aims to standardize the format and usage of sym files in order to avoid this problem.
Modifications, etc. should be suggested by [opening an issue](https://github.com/gbdev/rgbds-www/issues) or [starting a discussion](https://github.com/gbdev/rgbds-www/discussions) on this website's source repository, whichever is more appropriate.

Extensions to this format are known to exist ([example](https://github.com/mattcurrie/mgbdis/#symbol-files)), but they tend to be specific to a certain use case, and are generally a backwards-compatible superset of what are described below.

> Keep in mind that some sym files are written manually, such as when disassembling ROMs, so it is recommended for consumers to be lenient in what they accept.

Author: You're right, I'll remove this.
